content/recipes/interoperability/Storing text data in a knowledge graph.md

```diff
@@ -0,0 +1,306 @@
+    data = pd.DataFrame(requests.get(url).json().get('response').get('docs'))
+    # Return the first id
+    return data["iri"][0]
+  except:
+    return 'id-less'
+
+#create the graph
+g=Graph()
+EX = Namespace('http://example.org./')
+
+#dataframe storing the trios for visual clarity
+relations = pd.DataFrame()
+
+for i, paper in enumerate(papers['PubmedArticle']):
+    print("{}) {}".format(i+1, paper['MedlineCitation']['Article']['ArticleTitle']))
+    #get abstract
+    abstract_text_json = json.dumps(papers['PubmedArticle'][i]['MedlineCitation']['Article']['Abstract']['AbstractText'])
+    abstract_text = ' '.join(json.loads(abstract_text_json))
+    
+    #take through coreference model
+    coref_text = coref(abstract_text)._.resolved_text
+    
+    #entity extraction and linking
+    doc = rel_ext(coref_text)
+    
+    #adding trios to the graph
+    for value, rel_dict in doc._.rel.items():
+      subject_iri = ncit_iri(rel_dict['head_span']['text'])
+      object_iri = ncit_iri(rel_dict['head_span']['text'])
+      if subject_iri != 'id=less':
+        subj = URIRef(subject_iri)
+      else:
+        subj = EX[rel_dict['head_span']['text']]
+      if object_iri != 'id=less':
+        obj = URIRef(object_iri)
+      else:
+        subj = EX[rel_dict['head_span']['text']]  
+      pred = EX[Capitalise_underscore(rel_dict['relation'])]
+      g.add((subj, pred, obj))        
+    df = pd.DataFrame.from_dict(doc._.rel).transpose()
+    df['subject_text'] = df.head_span.apply(lambda x: x['text'])
+    df['subject_id'] = df.head_span.apply(lambda x: x['id'])
+    df['object_text'] = df.tail_span.apply(lambda x: x['text'])
+    df['object_id'] = df.tail_span.apply(lambda x: x['id'])
+
+    df = df.drop(["head_span", "tail_span"], axis = 1)
+
+    relations = pd.concat([relations, df])
+```
+
+Finally, we can visualize the resulting graph and export it in .ttl format.
+
+```python
+g.serialize(format = 'ttl')
+```
+
+```python
+import networkx as ntx
+import matplotlib.pyplot as plot
+
+graph = ntx.from_pandas_edgelist(relations, "subject_text", "object_text", edge_attr=True, create_using=ntx.MultiDiGraph())
+
+plot.figure(figsize=(10, 10))
+posn = ntx.spring_layout(graph)
+ntx.draw(graph, with_labels=True, node_color='green', edge_cmap=plot.cm.Blues, pos = posn)
+plot.show()
+```
+
+````{dropdown} 
+:open:
+```{figure} storing-text-data-in-KG-final-graph.png
+---
+height: 1000px
+name: Final knowledge graph visualiyation. The relations names are omitted.
+alt: Final knowledge graph visualiyation. The relations names are omitted.
+---
+Storing text data in the knowledge graph.
+```
+````
+
+## Conclusion
+Modern AI approaches allow us to traverse big chunks of unstructured text, extract information and order it in a form of a knowledge graph, making it FAIR. The details of the pipeline would vary depending on the specific use case, but the final result would definitely make your life easier!

```

Comment:
```suggestion
Modern AI/ML approaches allow to process large bodies of unstructured text, extract information and organise it in a form of knowledge graphs, in linked open data (LOD) graphs in ideal cases.
The present document provides a typical text processing pipeline to achieve this task. Of course, this is a simplified view and processing real life, large text corpora will require more advanced techniques to be deployed, from model training to development of specific algorithm to inclusion of expert curators to assist in the creation of the final knowledge graphs.
Still, we hope this will provide our readership with a basic understanding of the techniques available to move from unstructured text to generating a knowledge graph.



```